#! /usr/bin/env ruby

require 'uber-s3'
require 'json'
require 'yaml'
require 'optparse'

Environment = 'development'

Optparse = OptionParser.new do |opts|
  opts.banner = "Usage: s3_poller"

  opts.on_tail('-h', '--help', 'Print this help message') do
    puts opts
    exit
  end
  opts.on('-e ENVIRONMENT', 
          "Specify this application's environment") do |e|
    Environment = e
  end
end
Optparse.parse!

settings = YAML.load_file("./config/config.yml")[Environment]
Access_key = settings['aws_access_key_id']
Secret_access_key = settings['aws_secret_access_key']
Bucket = settings['aws_bucket']
Platform_support = "/platform-support"

# Connect to S3

S3 = UberS3.new({
                  :access_key => Access_key,
                  :secret_access_key => Secret_access_key,
                  :bucket => Bucket,
                  :adapter => :net_http
                })

class ::Hash
  # Define a deep merge for nested hashes
  def deep_merge(other)
    result = dup
    other.keys.each do |key|
      result[key] = if self[key].is_a? Hash and other[key].is_a? Hash
                      result[key].deep_merge(other[key])
                    else
                      other[key]
                    end
    end
    result
  end
end

def get_artifacts
  result = Hash.new
  begin
    # S3.objects(Platform_support).each was acting oddly, so I used this to work around it.
    S3.objects(Platform_support).to_a[1..-1].each do |support|
      hash = JSON.parse(support.value)
      result = result.deep_merge(hash)
    end
    result
  rescue UberS3::Error::Standard => e
    puts e.message
    puts <<EOF
[ERROR] Problem connecting to S3 using these credentials:
            aws access key id:      #{Access_key}
            aws secret access key:  #{Secret_access_key}
            bucket:                 #{Bucket}

Please check your credentials.
EOF
    exit
  end
end

build_list_path = settings['build_list']

File.open(build_list_path, "w") do |f|
  f.puts JSON.pretty_generate(get_artifacts)
end
